<a href="https://colab.research.google.com/github/GeneralMK7/ML-Internship/blob/main/task1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 📋 Task 1 : Understanding Dataset & Data Types

## 🎯 Objective
Understand the structure, data types, and ML readiness of the Titanic dataset.

## **Importing Required modules - numpy and pandas**

In [34]:
import numpy as np
import pandas as pd

## 1. **Loaded the Dataset**
- Used Pandas to load `train.csv`
- Displayed first and last few records using `head()` and `tail()`

In [35]:
train_dataset = pd.read_csv('Titanic_dataset/train.csv')

In [36]:
train_dataset.head() # if we send a value inside, it will return the no of rows from first to required.

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [37]:
train_dataset.tail() # if we send a value inside, it will return the no of rows from last to required

PassengerId  Survived  Pclass                                      Name  \
886          887         0       2                     Montvila, Rev. Juozas   
887          888         1       1              Graham, Miss. Margaret Edith   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch      Ticket   Fare Cabin Embarked  
886    male  27.0      0      0      211536  13.00   NaN        S  
887  female  19.0      0      0      112053  30.00   B42        S  
888  female   NaN      1      2  W./C. 6607  23.45   NaN        S  
889    male  26.0      0      0      111369  30.00  C148        C  
890    male  32.0      0      0      370376   7.75   NaN        Q

## 2. **Identified Data Types**
### Titanic Dataset – Columns

**PassengerId** – Numerical – Unique ID for each passenger

**Survived** – Binary – **0** = Did not survive, **1** = Survived (**TARGET VARIABLE**)

**Pclass** – Ordinal – Ticket class **(1 = First, 2 = Second, 3 = Third)**

**Name** – Categorical (Text) – Passenger Full Name

**Sex** – Categorical – Gender **(male, female)**

**Age** – Numerical **(Continuous)**– Age of the passenger in years

**SibSp** – Numerical **(Discrete)** – Number of siblings/spouses aboard

**Parch** – Numerical **(Discrete)** – Number of parents/children aboard

**Ticket** – Categorical – Ticket number/code

**Fare** – Numerical **(Continuous)** – Price paid for the ticket

**Cabin** – Categorical – Cabin number

**Embarked** – Categorical **(Nominal)** – Port of embarkation **(C = Cherbourg, Q = Queenstown, S = Southampton)**


In [43]:
train_dataset

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

## 3. **Analyzed Dataset**
- Used `df.info()` to check data types and null values
- Used `df.describe()` for statistical summary

In [41]:
train_dataset.dtypes # can also use this if you want nice format of types

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

In [42]:
train_dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [44]:
train_dataset.describe() # prints count,mean,standard deviation,min,max,etc..

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [45]:
train_dataset.isnull() # Returns a DataFrame of the same shape with True where values are missing and False otherwise.

PassengerId  Survived  Pclass   Name    Sex    Age  SibSp  Parch  Ticket  \
0          False     False   False  False  False  False  False  False   False   
1          False     False   False  False  False  False  False  False   False   
2          False     False   False  False  False  False  False  False   False   
3          False     False   False  False  False  False  False  False   False   
4          False     False   False  False  False  False  False  False   False   
..           ...       ...     ...    ...    ...    ...    ...    ...     ...   
886        False     False   False  False  False  False  False  False   False   
887        False     False   False  False  False  False  False  False   False   
888        False     False   False  False  False   True  False  False   False   
889        False     False   False  False  False  False  False  False   False   
890        False     False   False  False  False  False  False  False   False   

      Fare  Cabin  Embarked  
0    False   True     False  
1    False  False     False  
2    False   True     False  
3    False  False     False  
4    False   True     False  
..     ...    ...       ...  
886  False   True     False  
887  False  False     False  
888  False   True     False  
889  False  False     False  
890  False   True     False  

[891 rows x 12 columns]

In [51]:
train_dataset.isnull().sum() # Returns missing values per column.


PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [52]:
train_dataset.notnull().sum() # Returns non-null no of values

PassengerId    891
Survived       891
Pclass         891
Name           891
Sex            891
Age            714
SibSp          891
Parch          891
Ticket         891
Fare           891
Cabin          204
Embarked       889
dtype: int64

## 4. **Categorical Data Distribution Analysis**
- Analyzed unique values in all categorical columns.
- Identified the number of distinct categories in each feature.
- Observed binary and nominal categories such as Sex and Embarked.


In [65]:
train_dataset['Sex'].unique() # Checks Unique Values in Categorical Columns

array(['male', 'female'], dtype=object)

In [64]:
train_dataset['SibSp'].unique()

array([1, 0, 3, 4, 2, 5, 8])

In [66]:
train_dataset['Parch'].unique()

array([0, 1, 2, 5, 3, 4, 6])

In [67]:
train_dataset['Fare'].unique()

array([  7.25  ,  71.2833,   7.925 ,  53.1   ,   8.05  ,   8.4583,
        51.8625,  21.075 ,  11.1333,  30.0708,  16.7   ,  26.55  ,
        31.275 ,   7.8542,  16.    ,  29.125 ,  13.    ,  18.    ,
         7.225 ,  26.    ,   8.0292,  35.5   ,  31.3875, 263.    ,
         7.8792,   7.8958,  27.7208, 146.5208,   7.75  ,  10.5   ,
        82.1708,  52.    ,   7.2292,  11.2417,   9.475 ,  21.    ,
        41.5792,  15.5   ,  21.6792,  17.8   ,  39.6875,   7.8   ,
        76.7292,  61.9792,  27.75  ,  46.9   ,  80.    ,  83.475 ,
        27.9   ,  15.2458,   8.1583,   8.6625,  73.5   ,  14.4542,
        56.4958,   7.65  ,  29.    ,  12.475 ,   9.    ,   9.5   ,
         7.7875,  47.1   ,  15.85  ,  34.375 ,  61.175 ,  20.575 ,
        34.6542,  63.3583,  23.    ,  77.2875,   8.6542,   7.775 ,
        24.15  ,   9.825 ,  14.4583, 247.5208,   7.1417,  22.3583,
         6.975 ,   7.05  ,  14.5   ,  15.0458,  26.2833,   9.2167,
        79.2   ,   6.75  ,  11.5   ,  36.75  ,   7.7958,  12.5

In [68]:
train_dataset['Embarked'].unique()

array(['S', 'C', 'Q', nan], dtype=object)

In [60]:
categorical_cols = ['Name', 'Sex', 'Ticket', 'Cabin', 'Embarked']
for col in categorical_cols:
    print(f"\nColumn: {col}")
    print("Unique values:", train_dataset[col].nunique())
    print(train_dataset[col].value_counts())
    print('-------------------------------------------------------------')



Column: Name
Unique values: 891
Name
Dooley, Mr. Patrick                                    1
Braund, Mr. Owen Harris                                1
Cumings, Mrs. John Bradley (Florence Briggs Thayer)    1
Heikkinen, Miss. Laina                                 1
Futrelle, Mrs. Jacques Heath (Lily May Peel)           1
                                                      ..
Hewlett, Mrs. (Mary D Kingcome)                        1
Vestrom, Miss. Hulda Amanda Adolfina                   1
Andersson, Mr. Anders Johan                            1
Saundercock, Mr. William Henry                         1
Bonnell, Miss. Elizabeth                               1
Name: count, Length: 891, dtype: int64
-------------------------------------------------------------

Column: Sex
Unique values: 2
Sex
male      577
female    314
Name: count, dtype: int64
-------------------------------------------------------------

Column: Ticket
Unique values: 681
Ticket
347082              7
1601                

## 5. **Identifying Target Variable and Input Features**

- The target variable is the column that the model aims to predict.
- In the Titanic dataset, **Survived** is the target variable.
- It represents a binary classification problem (0 = Did not survive, 1 = Survived).
- Input features are the columns used to make predictions.
- Features such as **Age, Sex, Pclass, Fare, SibSp, Parch, and Embarked** are suitable for modeling.
- Identifier-like columns such as **PassengerId** and **Ticket** are not useful for prediction.


In [75]:
#If we want to predict the survival based on the dataset we can deop 'PassengerId' and 'ticket'
train_dataset_filter = train_dataset.drop(columns=['PassengerId', 'Ticket']) # Once run,we cannot take back the data if put into same variable

In [76]:
input_features = train_dataset_filter.drop(columns='Survived')
input_features

Pclass                                               Name     Sex   Age  \
0         3                            Braund, Mr. Owen Harris    male  22.0   
1         1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
2         3                             Heikkinen, Miss. Laina  female  26.0   
3         1       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
4         3                           Allen, Mr. William Henry    male  35.0   
..      ...                                                ...     ...   ...   
886       2                              Montvila, Rev. Juozas    male  27.0   
887       1                       Graham, Miss. Margaret Edith  female  19.0   
888       3           Johnston, Miss. Catherine Helen "Carrie"  female   NaN   
889       1                              Behr, Mr. Karl Howell    male  26.0   
890       3                                Dooley, Mr. Patrick    male  32.0   

     SibSp  Parch     Fare Cabin Embarked  
0        1      0   7.2500   NaN        S  
1        1      0  71.2833   C85        C  
2        0      0   7.9250   NaN        S  
3        1      0  53.1000  C123        S  
4        0      0   8.0500   NaN        S  
..     ...    ...      ...   ...      ...  
886      0      0  13.0000   NaN        S  
887      0      0  30.0000   B42        S  
888      1      2  23.4500   NaN        S  
889      0      0  30.0000  C148        C  
890      0      0   7.7500   NaN        Q  

[891 rows x 9 columns]

In [77]:
target_features = train_dataset_filter['Survived']
target_features

0      0
1      1
2      1
3      1
4      0
      ..
886    0
887    1
888    0
889    1
890    0
Name: Survived, Length: 891, dtype: int64

## 6. **Dataset Size and Machine Learning Suitability**

- The Titanic dataset contains 891 rows and 12 columns.
- This is considered a small dataset but is sufficient for basic machine learning tasks.
- The dataset has a clear target variable and meaningful input features.
- It represents a real-world problem, making it suitable for classification models.
- Small datasets may overfit, so careful validation is required.


In [80]:
train_dataset.shape # Prints the No. of rows and No. of columns
# In order to obtain more accuracy we need to get larger datasets

(891, 12)

## 7. **Data Quality Observations**

- Missing values are present in important columns such as Age(20%), Cabin (70 %),and Embarked.
- The Cabin column has a very high percentage of missing data, while Age has moderate missing values.
- The target variable (Survived) is imbalanced, with more passengers not surviving than surviving.
- Some columns such as PassengerId, Ticket, and Name behave like identifiers and do not contribute directly to prediction.
- These data quality issues must be handled through preprocessing before training machine learning models.


In [82]:
train_dataset.isnull().sum()


PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [83]:
(train_dataset.isnull().sum() / len(train_dataset)) * 100


PassengerId     0.000000
Survived        0.000000
Pclass          0.000000
Name            0.000000
Sex             0.000000
Age            19.865320
SibSp           0.000000
Parch           0.000000
Ticket          0.000000
Fare            0.000000
Cabin          77.104377
Embarked        0.224467
dtype: float64

In [85]:
train_dataset['Survived'].value_counts() # Checks class imbalance


Survived
0    549
1    342
Name: count, dtype: int64